internal/tiltfile2/tiltfile_state.go

```diff
@@ -0,0 +1,292 @@
+	"github.com/google/skylark"
+
+	"github.com/windmilleng/tilt/internal/k8s"
+	"github.com/windmilleng/tilt/internal/logger"
+	"github.com/windmilleng/tilt/internal/model"
+)
+
+type tiltfileState struct {
+	// set at creation
+	ctx      context.Context
+	filename string
+
+	// added to during execution
+	configFiles    []string
+	images         []*dockerImage
+	imagesByName   map[string]*dockerImage
+	k8s            []*k8sResource
+	k8sByName      map[string]*k8sResource
+	k8sUnresourced []k8s.K8sEntity
+
+	// for assembly
+	usedImages map[string]bool
+}
+
+func newTiltfileState(ctx context.Context, filename string) *tiltfileState {
+	return &tiltfileState{
+		ctx:          ctx,
+		filename:     filename,
+		imagesByName: make(map[string]*dockerImage),
+		k8sByName:    make(map[string]*k8sResource),
+		configFiles:  []string{filename},
+		usedImages:   make(map[string]bool),
+	}
+}
+
+func (s *tiltfileState) exec() error {
+	thread := &skylark.Thread{
+		Print: func(_ *skylark.Thread, msg string) {
+			logger.Get(s.ctx).Infof("%s", msg)
+		},
+	}
+	_, err := skylark.ExecFile(thread, s.filename, nil, s.builtins())
+	return err
+}
+
+// Builtin functions
+
+const (
+	// build functions
+	dockerBuildN = "docker_build"
+	fastBuildN   = "fast_build"
+
+	// k8s functions
+	k8sYamlN     = "k8s_yaml"
+	k8sResourceN = "k8s_resource"
+	portForwardN = "port_forward"
+
+	// file functions
+	localGitRepoN = "local_git_repo"
+	localN        = "local"
+	readFileN     = "read_file"
+	kustomizeN    = "kustomize"
+)
+
+func (s *tiltfileState) builtins() skylark.StringDict {
+	r := make(skylark.StringDict)
+	add := func(name string, fn func(thread *skylark.Thread, fn *skylark.Builtin, args skylark.Tuple, kwargs []skylark.Tuple) (skylark.Value, error)) {
+		r[name] = skylark.NewBuiltin(name, fn)
+	}
+
+	add(dockerBuildN, s.dockerBuild)
+	add(fastBuildN, s.fastBuild)
+
+	add(k8sYamlN, s.k8sYaml)
+	add(k8sResourceN, s.k8sResource)
+	add(portForwardN, s.portForward)
+
+	add(localGitRepoN, s.localGitRepo)
+	add(localN, s.local)
+	add(readFileN, s.skylarkReadFile)
+	add(kustomizeN, s.kustomize)
+	return r
+}
+
+const unresourcedName = "k8s_yaml"
+
+func (s *tiltfileState) assemble() ([]*k8sResource, error) {
+	hadUnresourced := len(s.k8sUnresourced) > 0
+
+	images, err := s.findUnresourcedImages()
+	if err != nil {
+		return nil, err
+	}
+	for _, image := range images {
+		if _, ok := s.imagesByName[image.Name()]; !ok {
+			// only expand for images we know how to build
+			continue
+		}
+		target, err := s.findExpandTarget(image)
+		if err != nil {
+			return nil, err
+		}
+		if err := s.extractImage(target, image); err != nil {
+			return nil, err
+		}
+	}
+
+	if hadUnresourced {
+		r, err := s.makeK8sResource(unresourcedName)
+		if err != nil {
+			return nil, err
+		}
+		r.k8s = s.k8sUnresourced
+	}
+
+	assembledImages := map[string]bool{}
+	for _, r := range s.k8s {
+		if err := s.validateK8s(r, assembledImages); err != nil {
```

Comment:
why does this function add to the map as a side effect? seems more readable to check the err and, if no err, add to the map